src/devices/secplus_v2.c

```diff
@@ -0,0 +1,444 @@
+        x >>= 1;
+    }
+
+    // fprintf(stderr, "f1 (%d) %d %d %d\n", part_id, p0, p1, p2);
+
+    // selectively invert buffers
+    switch (invert) {
+    case 0x00: // 0b0000 (True, True, False),
+        p0 = ~p0 & 0x03FF;
+        p1 = ~p1 & 0x03FF;
+        break;
+    case 0x01: // 0b0001 (False, True, False),
+        p1 = ~p1 & 0x03FF;
+        break;
+    case 0x02: // 0b0010 (False, False, True),
+        p2 = ~p2 & 0x03FF;
+        break;
+    case 0x04: // 0b0100 (True, True, True),
+        p0 = ~p0 & 0x03FF;
+        p1 = ~p1 & 0x03FF;
+        p2 = ~p2 & 0x03FF;
+        break;
+    case 0x05: // 0b0101 (True, False, True),
+    case 0x0a: // 0b1010 (True, False, True),
+        p0 = ~p0 & 0x03FF;
+        p2 = ~p2 & 0x03FF;
+        break;
+    case 0x06: // 0b0110 (False, True, True),
+        p1 = ~p1 & 0x03FF;
+        p2 = ~p2 & 0x03FF;
+        break;
+    case 0x08: // 0b1000 (True, False, False),
+        p0 = ~p0 & 0x03FF;
+        break;
+    case 0x09: // 0b1001 (False, False, False),
+        break;
+    default:
+        if (verbose) 
+            fprintf(stderr, "Invert FAIL\n");
+        return 1;
+    }
+
+    uint16_t a = p0, b = p1, c = p2;
+
+    // selectively reorder buffers
+    switch (order) {
+    case 0x06: // 0b0110  2, 1, 0],
+    case 0x09: // 0b1001  2, 1, 0],
+        p2 = a;
+        p1 = b;
+        p0 = c;
+        break;
+
+    case 0x08: // 0b1000  1, 2, 0],
+    case 0x04: // 0b0100  1, 2, 0],
+        p1 = a;
+        p2 = b;
+        p0 = c;
+        break;
+
+    case 0x01: // 0b0001 2, 0, 1],
+        p2 = a;
+        p0 = b;
+        p1 = c;
+        break;
+
+    case 0x00: // 0b0000  0, 2, 1],
+        p0 = a;
+        p2 = b;
+        p1 = c;
+        break;
+
+    case 0x05: // 0b0101 1, 0, 2],
+        p1 = a;
+        p0 = b;
+        p2 = c;
+        break;
+
+    case 0x02: // 0b0010 0, 1, 2],
+    case 0x0A: // 0b1010 0, 1, 2],
+        p0 = a;
+        p1 = b;
+        p2 = c;
+        break;
+
+    default:
+        if (verbose)
+            fprintf(stderr, "Order FAIL");
+        return 2;
+    }
+
+    bitbuffer_extract_bytes(bits, 0, 4, buffy, 8);
+    x     = buffy[0];
+    int k = 0;
+    for (int i = 6; i >= 0; i -= 2) {
+        roll_array[k++] = (x >> i) & 0x03;
+    }
+
+    // bitrow_printf(buffy, 8, "%s ; buffy bits ", __func__);
+
+    // assemble binary bits into trinary 
+    x = p2;
+    for (int i = 8; i >= 0; i -= 2) {
+        roll_array[k++] = (x >> i) & 0x03;
+    }
+
+    if (verbose) {
+        fprintf(stderr, "%s : roll_array : (%d) ", __func__, part_id);
+        for (int i = 0; i < 9; i++) {
+            fprintf(stderr, "%d ", roll_array[i]);
+        }
+        fprintf(stderr, "\n");
+    }
+
+    // SANITY check trinary valuse, 00/01/10 are valid,  11 is not 
+    for (int i = 0; i < 9; i++) {
+        if (roll_array[i] == 3) {
+            fprintf(stderr, "roll_array val  FAIL\n");
+            return 1; // DECODE_FAIL_SANITY;
```

Comment:
You can just return DECODE_FAIL_SANITY here, nothing is done with the return value anyway ;)